p2p/ipld/write.go

```diff
@@ -52,113 +47,70 @@ func PutBlock(
 	if err != nil {
 		return fmt.Errorf("failure to recompute the extended data square: %w", err)
 	}
-	// get row and col roots to be provided
-	// this also triggers adding data to DAG
-	prov := newProvider(ctx, croute, int32(squareSize*4), logger.With("height", block.Height))
-	for _, root := range eds.RowRoots() {
-		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
-	}
-	for _, root := range eds.ColumnRoots() {
-		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
+
+	// setup provide workers
+	const workers = 32
+	var wg stdsync.WaitGroup
+	jobs := make(chan []byte, workers)
+	errc := make(chan error, 1)
+	workerCtx, workerCancel := context.WithCancel(ctx)
+	defer workerCancel() // ensure that this cancel is called
+
+	// start feeding the workers jobs
+	go feedJobs(workerCtx, jobs, eds.RowRoots(), eds.ColumnRoots())
+
+	// start each worker
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		worker := provideWorker{croute, logger, workerCancel}
+		go worker.provide(workerCtx, &wg, jobs, errc)
 	}
-	// commit the batch to ipfs
-	err = batchAdder.Commit()
+
+	// for for all jobs to finish, an error to occur, or a if a timeout is reached
+	wg.Wait()
+	err = collectErrors(errc)
 	if err != nil {
 		return err
 	}
-	// wait until we provided all the roots if requested
-	<-prov.Done()
-	return prov.Err()
-}
-
-var provideWorkers = 32
-
-type provider struct {
-	ctx  context.Context
-	done chan struct{}
-
-	err   error
-	errLk sync.RWMutex
-
-	jobs  chan cid.Cid
-	total int32
-
-	croute routing.ContentRouting
-	log    log.Logger
-	took   time.Time
-}
 
-func newProvider(ctx context.Context, croute routing.ContentRouting, toProvide int32, logger log.Logger) *provider {
-	p := &provider{
-		ctx:    ctx,
-		done:   make(chan struct{}),
-		jobs:   make(chan cid.Cid, provideWorkers),
-		total:  toProvide,
-		croute: croute,
-		log:    logger,
-	}
-	for range make([]bool, provideWorkers) {
-		go p.worker()
-	}
-	logger.Info("Started Providing to DHT")
-	p.took = time.Now()
-	return p
-}
-
-func (p *provider) Provide(id cid.Cid) {
-	select {
-	case p.jobs <- id:
-	case <-p.ctx.Done():
-	}
```

Comment:
BTW, just to be clear: I do feel that this code is more readable and accessible. We can followup on this in the next iteration for PutBlock.